scripts/inject-document-props-as-overload.py

```diff
@@ -0,0 +1,87 @@
+import inspect
+import re
+import warnings
+from operator import itemgetter
+from typing import Optional, Tuple, List
+
+from jina import Document
+
+
+def get_properties(cls) -> List[Tuple[str, Optional[str], Optional[str]]]:
+    src = inspect.getsource(cls)
+    members = dict(inspect.getmembers(cls))
+    setters = re.findall(
+        r'@[a-zA-Z0-9_]+\.setter\s+def\s+([a-zA-Z0-9_]+)\s*\(self,\s*[a-zA-Z0-9_]+\s*:\s*(.*?)\)',
+        src,
+        flags=re.DOTALL,
+    )
+
+    property_docs = []
+    for setter, _ in setters:
+        if setter not in members:
+            warnings.warn(
+                f'{setter} is found as a setter but there is no corresponding getter'
+            )
+            property_docs.append(None)
+        else:
+            doc = inspect.getdoc(members[setter])
+            description = next(iter(re.findall(':return:(.*)', doc)), None)
+            if description:
+                description = description.strip()
+            property_docs.append(description)
+    return list(
+        zip(map(itemgetter(0), setters), map(itemgetter(1), setters), property_docs)
+    )
+
+
+def get_overload_signature(
+    properties,
+    indent=' ' * 4,
+):
+    kwargs = [
+        f'{indent}{indent}{property_name}: Optional[{type_hint}] = None'
+        for property_name, type_hint, _ in properties
+    ]
+    args_str = ', \n'.join(kwargs + [f'{indent}{indent}**kwargs'])
+    doc_str = '\n'.join(
+        f'{indent}{indent}:param {property_name}: {description}'
+        for property_name, _, description in properties
+    )
+
+    noqa_str = '\n'.join(
+        f'{indent}{indent}.. # noqa: DAR{j}' for j in ['202', '101', '003']
```

Comment:
what is this `noqa` doing? It is weird to rely on this being present right?

Reply:
you're right
I just wanted to be extra safe but I don't think we'll ever be in a case of tolerating these darglint errors